components/tidb_query_codegen/src/rpn_function.rs

```diff
@@ -475,12 +492,67 @@ impl ValidatorFnGenerator {
     }
 }
 
+fn generate_init_metadata_fn(
+    metadata_ctor: &Option<TokenStream>,
+    impl_generics: &ImplGenerics<'_>,
+    where_clause: Option<&WhereClause>,
+) -> TokenStream {
+    let fn_body = if let Some(metadata_ctor) = metadata_ctor {
+        quote! { Box::new(#metadata_ctor(expr)) }
+    } else {
+        quote! { Box::new(()) }
+    };
+    quote! {
+        fn init_metadata #impl_generics (expr: &mut ::tipb::Expr)
+            -> Box<dyn std::any::Any + Send> #where_clause {
+            #fn_body
+        }
+    }
+}
+
+fn generate_downcast_metadata(has_metadata: bool) -> TokenStream {
+    if has_metadata {
+        quote! {
+            let metadata = std::any::Any::downcast_ref(metadata).expect("downcast metadata error");
+        }
+    } else {
+        quote! {}
+    }
+}
+
+fn generate_metadata_type_checker(
+    metadata_ctor: &Option<TokenStream>,
+    impl_generics: &ImplGenerics<'_>,
+    where_clause: Option<&WhereClause>,
+    fn_body: impl FnOnce(&TokenStream) -> TokenStream,
+) -> TokenStream {
+    if let Some(metadata_ctor) = metadata_ctor {
+        let body = fn_body(metadata_ctor);
```

Comment:
Could we move the `if let` and `fn_body` expansion to the callee so as to eliminate this confusing closure?

Reply:
I extracted them to avoid repeating `if ... { quote! { ...} } else { quote! {} }`.

Reply:
The closure is natural because the body is to fit the corresponding signature of the rpn function.

Reply:
I got it. Never mind.

Reply:
I got it. Never mind.